File: app/src/main/java/com/example/dagger_hilt/sys/util/Resource.java
package com.example.dagger_hilt.sys.util;

import androidx.annotation.Nullable;

public record Resource<T>(Constants.StatusResponse status, @Nullable T data, String message) {

    public static <T> Resource<T> success(T data) {
        return new Resource<>(Constants.StatusResponse.SUCCESS, data, null);
    }

    public static <T> Resource<T> error(String msg, @Nullable T data) {
        return new Resource<>(Constants.StatusResponse.ERROR, data, msg);
    }

    public static <T> Resource<T> loading(@Nullable T data) {
        return new Resource<>(Constants.StatusResponse.LOADING, data, null);
    }
}

File: app/src/main/java/com/example/dagger_hilt/sys/util/Resource.kt
package com.example.dagger_hilt.sys.util

data class Resource<out T>(
    val status: Constants.StatusResponse,
    val data: T?,
    val message: String?
) {
    companion object {
        fun <T> success(data: T?): Resource<T> =
            Resource(Constants.StatusResponse.SUCCESS, data, null)

        fun <T> error(msg: String): Resource<T> =
            Resource(Constants.StatusResponse.ERROR, null, msg)

        fun <T> loading(): Resource<T> = Resource(Constants.StatusResponse.LOADING, null, null)

        fun <T> blank(): Resource<T> = Resource(Constants.StatusResponse.BLANK, null, null)
    }
}
